File: src/lib/foncierRAG.ts
// Moteur RAG 100% local pour Fitila Tem IA
// Aucun appel réseau — BM25 + bonus n-grammes + extraction de phrases.
// Corpus : 207 articles du Code Foncier béninois en Bariba.

import { FONCIER_CORPUS, type FoncierArticle } from '@/data/foncierBaribaCorpus';

// ──────────────────────────────────────────────────────────────────────
// Stop-words FR + BA fréquents (à ignorer dans le scoring)
// ──────────────────────────────────────────────────────────────────────
const STOP_WORDS = new Set([
  // FR
  'le','la','les','un','une','des','de','du','et','ou','est','sont','ce','cette','ces','que','qui','quoi','quel','quelle',
  'pour','par','sur','dans','avec','sans','au','aux','en','à','a','se','sa','son','ses','mon','ma','mes','ton','ta','tes',
  'il','elle','ils','elles','je','tu','nous','vous','on','y','ne','pas','plus','être','avoir','faire','dit','dire',
  // BA basiques (mots-outils très fréquents — à raffiner empiriquement)
  'ka','ya','ye','yè','yi','yì','ba','bà','bù','bu','wã','wãa','sɔɔ','sɔ̃','mi','mɛ','mɛ̃','tè','tì','ta','sere','ǹ','kun','tera','yera','ma','ko','koo','mɔ','mɔ̀','sãa','sãawa','niya','nu','nì','sin','si','sì','su','goo','geja',
]);

// ──────────────────────────────────────────────────────────────────────
// Tokenizer Bariba-aware
// ──────────────────────────────────────────────────────────────────────
function normalize(text: string): string {
  return text.normalize('NFC').toLowerCase();
}

function tokenize(text: string): string[] {
  const norm = normalize(text);
  // Garde lettres latines + caractères Bariba (ɔ ɛ ã ĩ ũ ɔ̃ ɛ̃ ǹ etc.) + chiffres
  const cleaned = norm.replace(/[^\p{L}\p{M}\p{N}\s]/gu, ' ');
  return cleaned
    .split(/\s+/)
    .filter(t => t.length >= 2 && !STOP_WORDS.has(t));
}

function bigrams(tokens: string[]): string[] {
  const out: string[] = [];
  for (let i = 0; i < tokens.length - 1; i++) {
    out.push(tokens[i] + ' ' + tokens[i + 1]);
  }
  return out;
}

// ──────────────────────────────────────────────────────────────────────
// Pré-construction de l'index BM25 (une seule fois au load du module)
// ──────────────────────────────────────────────────────────────────────
interface IndexedDoc {
  article: FoncierArticle;
  tokens: string[];
  tf: Map<string, number>;
  bigrams: Set<string>;
  length: number;
}

const K1 = 1.5;
const B = 0.75;

function buildIndex() {
  const docs: IndexedDoc[] = FONCIER_CORPUS.map(article => {
    const fullText = [article.number, article.gariWiru, article.bonu, article.content]
      .filter(Boolean).join(' ');
    const tokens = tokenize(fullText);
    const tf = new Map<string, number>();
    for (const t of tokens) tf.set(t, (tf.get(t) || 0) + 1);
    return {
      article,
      tokens,
      tf,
      bigrams: new Set(bigrams(tokens)),
      length: tokens.length,
    };
  });

  const df = new Map<string, number>();
  for (const d of docs) {
    for (const term of d.tf.keys()) {
      df.set(term, (df.get(term) || 0) + 1);
    }
  }

  const N = docs.length;
  const avgdl = docs.reduce((s, d) => s + d.length, 0) / N;
  const idf = new Map<string, number>();
  for (const [term, freq] of df) {
    idf.set(term, Math.log(1 + (N - freq + 0.5) / (freq + 0.5)));
  }

  return { docs, idf, avgdl };
}

const INDEX = buildIndex();

// ──────────────────────────────────────────────────────────────────────
// Recherche
// ──────────────────────────────────────────────────────────────────────
export interface SearchHit {
  article: FoncierArticle;
  score: number;
}

function detectArticleMention(query: string): number | null {
  // "saria 14", "saria 14se", "article 14", "art 14"
  const m = normalize(query).match(/\b(?:saria|article|art|art\.)\s*(\d{1,3})\b/);
  if (m) return parseInt(m[1], 10);
  return null;
}

export function searchFoncier(query: string, k = 6): SearchHit[] {
  const qTokens = tokenize(query);
  if (qTokens.length === 0) return [];
  const qBigrams = bigrams(qTokens);
  const mentioned = detectArticleMention(query);

  const hits: SearchHit[] = INDEX.docs.map(d => {
    let score = 0;
    for (const qt of qTokens) {
      const tf = d.tf.get(qt) || 0;
      if (tf === 0) continue;
      const idf = INDEX.idf.get(qt) || 0;
      const denom = tf + K1 * (1 - B + B * (d.length / INDEX.avgdl));
      score += idf * (tf * (K1 + 1)) / denom;
    }
    // Bonus bigrammes (expressions)
    for (const bg of qBigrams) {
      if (d.bigrams.has(bg)) score += 2.0;
    }
    // Bonus mention explicite d'un numéro de saria
    if (mentioned !== null && d.article.number.includes(`${mentioned}se`)) {
      score += 50;
    }
    if (mentioned !== null && d.article.id === mentioned) {
      score += 50;
    }
    return { article: d.article, score };
  });

  return hits
    .filter(h => h.score > 0)
    .sort((a, b) => b.score - a.score)
    .slice(0, k);
}

// ──────────────────────────────────────────────────────────────────────
// Extracteur de réponse (purement extractif)
// ──────────────────────────────────────────────────────────────────────
const FALLBACK_BA = 'Gari yini bweseru ku wáa tem saria tire teni søø.';
const SCORE_THRESHOLD = 1.2;

export interface AnswerResult {
  answer: string;
  sources: { id: number; number: string; content: string; page: number; gariWiru?: string | null }[];
  isFallback: boolean;
}

function splitSentences(text: string): string[] {
  return text
    .split(/(?<=[.!?])\s+|\n+/)
    .map(s => s.trim())
    .filter(s => s.length > 5);
}

function scoreSentence(sentence: string, qTokenSet: Set<string>): number {
  const tokens = tokenize(sentence);
  if (tokens.length === 0) return 0;
  let hits = 0;
  for (const t of tokens) if (qTokenSet.has(t)) hits++;
  // Score = nb de matches, légèrement normalisé pour éviter le bruit des phrases longues
  return hits + hits / Math.sqrt(tokens.length);
}

export function answerFromCorpus(query: string): AnswerResult {
  const hits = searchFoncier(query, 6);
  if (hits.length === 0 || hits[0].score < SCORE_THRESHOLD) {
    return { answer: FALLBACK_BA, sources: [], isFallback: true };
  }

  const qTokenSet = new Set(tokenize(query));

  // Récolte de phrases candidates depuis les top-K articles
  type Candidate = { sentence: string; score: number; article: FoncierArticle };
  const candidates: Candidate[] = [];
  for (const h of hits) {
    const sentences = splitSentences(h.article.content);
    for (const s of sentences) {
      const sc = scoreSentence(s, qTokenSet) * Math.log(1 + h.score);
      if (sc > 0) candidates.push({ sentence: s, score: sc, article: h.article });
    }
  }

  if (candidates.length === 0) {
    // Aucune phrase ne matche → on retourne le contenu intégral du meilleur article
    const best = hits[0].article;
    return {
      answer: best.content,
      sources: [{ id: best.id, number: best.number, content: best.content, page: best.page, gariWiru: best.gariWiru }],
      isFallback: false,
    };
  }

  candidates.sort((a, b) => b.score - a.score);

  // Top 3 phrases distinctes
  const seen = new Set<string>();
  const picked: Candidate[] = [];
  for (const c of candidates) {
    if (seen.has(c.sentence)) continue;
    seen.add(c.sentence);
    picked.push(c);
    if (picked.length >= 3) break;
  }

  // Sources uniques (préserve l'ordre d'apparition)
  const sourceMap = new Map<number, FoncierArticle>();
  for (const p of picked) {
    if (!sourceMap.has(p.article.id)) sourceMap.set(p.article.id, p.article);
  }

  const answer = picked.map(p => p.sentence).join('\n\n');
  const sources = Array.from(sourceMap.values()).map(a => ({
    id: a.id, number: a.number, content: a.content, page: a.page, gariWiru: a.gariWiru,
  }));

  return { answer, sources, isFallback: false };
}

File: fitila_flutter/lib/core/foncier_rag.dart
import 'dart:convert';
import 'dart:math' as math;

import 'package:flutter/services.dart';

class FoncierSource {
  const FoncierSource({
    required this.id,
    required this.number,
    required this.content,
    required this.page,
    this.book,
  });

  final int id;
  final String number;
  final String content;
  final int page;
  final String? book;
}

class FoncierAnswer {
  const FoncierAnswer({
    required this.answer,
    required this.sources,
    required this.isFallback,
  });

  final String answer;
  final List<FoncierSource> sources;
  final bool isFallback;
}

class _IndexedArticle {
  const _IndexedArticle({required this.source, required this.tokens});

  final FoncierSource source;
  final List<String> tokens;
}

class FoncierRag {
  static const _fallback = 'Gari yini bweseru ku wáa tem saria tire teni sɔɔ.';
  static const _stopWords = <String>{
    'le',
    'la',
    'les',
    'un',
    'une',
    'des',
    'de',
    'du',
    'et',
    'ou',
    'est',
    'sont',
    'ce',
    'cette',
    'ces',
    'que',
    'qui',
    'quoi',
    'pour',
    'par',
    'sur',
    'dans',
    'avec',
    'sans',
    'au',
    'aux',
    'en',
    'à',
    'a',
    'se',
    'sa',
    'son',
    'il',
    'elle',
    'je',
    'tu',
    'nous',
    'vous',
    'ka',
    'ya',
    'ye',
    'yè',
    'yi',
    'ba',
    'wã',
    'wãa',
    'sɔɔ',
    'sɔ̃',
    'mi',
    'mɛ',
    'ta',
    'sere',
    'kun',
    'ko',
  };

  static List<_IndexedArticle>? _index;

  static List<String> _tokens(String text) {
    final cleaned = text.toLowerCase().replaceAll(
      RegExp(r'''[.,;:!?()\[\]{}"“”'’/\\\-]'''),
      ' ',
    );
    return cleaned
        .split(RegExp(r'\s+'))
        .where((value) => value.length >= 2 && !_stopWords.contains(value))
        .toList(growable: false);
  }

  static Future<List<_IndexedArticle>> _load() async {
    if (_index != null) return _index!;
    final raw = await rootBundle.loadString(
      'assets/data/foncier_bariba_corpus.json',
    );
    final decoded = jsonDecode(raw) as List<dynamic>;
    _index = decoded
        .map((value) {
          final row = value as Map<String, dynamic>;
          final source = FoncierSource(
            id: (row['id'] as num?)?.toInt() ?? 0,
            number: row['number']?.toString() ?? '',
            content: row['content']?.toString() ?? '',
            page: (row['page'] as num?)?.toInt() ?? 0,
            book: row['gariWiru']?.toString(),
          );
          return _IndexedArticle(
            source: source,
            tokens: _tokens(
              '${source.number} ${source.book ?? ''} ${row['bonu'] ?? ''} ${source.content}',
            ),
          );
        })
        .toList(growable: false);
    return _index!;
  }

  static Future<FoncierAnswer> answer(String query) async {
    final queryTokens = _tokens(query);
    if (queryTokens.isEmpty) {
      return const FoncierAnswer(
        answer: _fallback,
        sources: [],
        isFallback: true,
      );
    }
    final articleMention = RegExp(
      r'\b(?:saria|article|art\.?)[ ]*(\d{1,3})(?:se)?\b',
      caseSensitive: false,
    ).firstMatch(query)?.group(1);
    final index = await _load();
    final frequencies = <String, int>{};
    for (final term in queryTokens) {
      frequencies[term] = index
          .where((doc) => doc.tokens.contains(term))
          .length;
    }

    final ranked = <(_IndexedArticle, double)>[];
    for (final document in index) {
      var score = 0.0;
      for (final term in queryTokens) {
        final count = document.tokens.where((token) => token == term).length;
        if (count == 0) continue;
        final idf = math.log(1 + index.length / (1 + (frequencies[term] ?? 0)));
        score += count * idf;
      }
      if (articleMention != null &&
          (document.source.id.toString() == articleMention ||
              document.source.number.contains('${articleMention}se'))) {
        score += 100;
      }
      if (score > 0) ranked.add((document, score));
    }
    ranked.sort((a, b) => b.$2.compareTo(a.$2));
    if (ranked.isEmpty || ranked.first.$2 < 1.2) {
      return const FoncierAnswer(
        answer: _fallback,
        sources: [],
        isFallback: true,
      );
    }

    if (articleMention != null) {
      final exact = ranked
          .map((hit) => hit.$1.source)
          .where(
            (source) =>
                source.id.toString() == articleMention ||
                source.number.contains('${articleMention}se'),
          )
          .firstOrNull;
      if (exact != null) {
        return FoncierAnswer(
          answer: exact.content,
          sources: [exact],
          isFallback: false,
        );
      }
    }

    final selected = ranked.take(3).map((hit) => hit.$1.source).toList();
    final candidates = <(String, double, FoncierSource)>[];
    final querySet = queryTokens.toSet();
    for (final source in selected) {
      final sentences = source.content
          .split(RegExp(r'(?<=[.!?])\s+|\n+'))
          .map((value) => value.trim())
          .where((value) => value.length > 5);
      for (final sentence in sentences) {
        final words = _tokens(sentence);
        final matches = words.where(querySet.contains).length;
        if (matches > 0) {
          candidates.add((
            sentence,
            matches + matches / math.sqrt(words.length),
            source,
          ));
        }
      }
    }
    candidates.sort((a, b) => b.$2.compareTo(a.$2));
    final picked = candidates.take(3).toList();
    final answer = picked.isEmpty
        ? selected.first.content
        : picked.map((item) => item.$1).join('\n\n');
    final sourceIds = picked.map((item) => item.$3.id).toSet();
    final sources = picked.isEmpty
        ? [selected.first]
        : selected.where((source) => sourceIds.contains(source.id)).toList();
    return FoncierAnswer(answer: answer, sources: sources, isFallback: false);
  }
}
